### PR TITLE
♻️ Polished and modernised agent service ⚠️

### PR DESCRIPTION
**Pull request by GitHK**
## Dev-ops ⚠️
- no action required

## What do these changes do?

This service is way out of date and did not receive attention in years (2+). It currently causes a lot of pain for the platform and with the latest tools we've added it can be rewritten in a more sane and efficient manner.

This rewrite is also keeping an eye out for cost reduction, tries to avoid useless data uploads (unlike the previous version).

Highlights:
- `agent` nodes are targeted via `RPCNamespace` which is composed out of `docker_node_id` + `swarm_stack_name`
- only one volume at a time can be removed per node (avoids race conditions as they were present in the previous version)
- `agent` still maintains the task of backing up data that was left on the node by sidecars non terminated by director-v2
- `agent` exposes Prometheus metrics regarding the number of volumes removed and the number of volumes that it backed up to S3
- dropped `dy-volrm` docker service which was created for each service for the purpose of removing it's volumes

Followup issues:
- https://github.com/ITISFoundation/osparc-simcore/issues/6469
- https://github.com/ITISFoundation/osparc-simcore/issues/3939

## Related issue/s

- https://github.com/ITISFoundation/osparc-issues/issues/638
- fixes https://github.com/ITISFoundation/osparc-simcore/issues/3446
- fixes https://github.com/ITISFoundation/osparc-issues/issues/763
- fixes https://github.com/ITISFoundation/osparc-simcore/issues/3890
- closes https://github.com/ITISFoundation/osparc-simcore/pull/3941

## Dev-ops checklist ⚠️

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))
- [x] [must merge this MR](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/merge_requests/954) before this PR

### CODE REVIEW

File: services/agent/src/simcore_service_agent/core/api/rpc/routes.py
Comment (pcrespov) [inline services/agent/src/simcore_service_agent/core/api/rpc/routes.py:1]: in fastapi services the api package including api.rest and api.rpc we normally have it at the same level as services
the idea is to highlight controler+service+repository layers by packages  named as `api`, `services`, `db`

File: services/agent/src/simcore_service_agent/core/_dependencies.py
Comment (pcrespov) [inline services/agent/src/simcore_service_agent/core/_dependencies.py:1]: Dependency inject can only be used in fastapi in rest handlers. That is the reason why we also add this module as protected **within** the `api.rest` packages.

File: services/agent/src/simcore_service_agent/services/models.py
Comment (pcrespov) [inline services/agent/src/simcore_service_agent/services/models.py:1]: Separate models from services. In fastapi libraries we put them in its own package `models`

File: services/agent/src/simcore_service_agent/services/volumes_manager.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+from ..core.settings import ApplicationSettings
+from .docker_utils import get_unused_dynamc_sidecar_volumes, remove_volume
+
+_logger = logging.getLogger(__name__)
+
+_WAIT_FOR_UNUSED_SERVICE_VOLUMES: Final[timedelta] = timedelta(minutes=1)
+
+
+@dataclass
+class VolumesManager:
```
Comment (pcrespov) [inline services/agent/src/simcore_service_agent/services/volumes_manager.py:26]: Why not use our friend SingletonInAppStateMixin ? This way setters and getters of app.state are all stardardized

File: services/agent/src/simcore_service_agent/core/api/rpc/_volumes.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+@router.expose()
+async def remove_volumes_without_backup_for_service(
+    app: FastAPI, *, node_id: NodeID
+) -> None:
+    with log_context(_logger, logging.INFO, f"removing volumes for service: {node_id}"):
+        await get_volumes_manager(app).remove_service_volumes(node_id)
+
+
+@router.expose()
```
Comment (pcrespov) [inline services/agent/src/simcore_service_agent/core/api/rpc/_volumes.py:22]: YOu do not expose any exceptions to the interface (see `reraise_if_error_type`)? how do you handle errors?

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events_utils.py [hunk 3]
```diff
@@ -230,29 +230,13 @@ async def service_remove_sidecar_proxy_docker_networks_and_volumes(
...
-                    scheduler_data.paths_mapping.outputs_path,
-                    *scheduler_data.paths_mapping.state_paths,
-                ]
-            ]
-            with log_context(
-                _logger, logging.DEBUG, f"removing volumes via service for {node_uuid}"
-            ):
+            with log_context(_logger, logging.DEBUG, f"removing volumes '{node_uuid}'"):
+                rabbit_rpc_client: RabbitMQRPCClient = app.state.rabbitmq_rpc_client
                 await remove_volumes_from_node(
```
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events_utils.py:235]: Q: so you ask the agent to remove the volumes from one of the nodes. Every node has one simcore agent service (was setup as `global`)  how do you know that the correct agent is taking this task?

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/volumes_removal.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+async def remove_volumes_from_node(
+    rabbit_rpc_client: RabbitMQRPCClient,
+    docker_node_id: str,
+    swarm_stack_name: str,
+    *,
+    node_id: NodeID,
+) -> None:
+    """removes all service volumes form the node where it was running"""
+
+    await remove_volumes_without_backup_for_service(
```
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/volumes_removal.py:17]: Q: what is the purpose of creating a full new module for a one-to-one function? 
NOte that here you can literaly make
```python

remove_volumes_from_node = remove_volumes_without_backup_for_service
```

File: services/agent/src/simcore_service_agent/core/api/rpc/_volumes.py (hunk 2 above)
Comment (pcrespov) [reply services/agent/src/simcore_service_agent/core/api/rpc/_volumes.py:22]: Perhaps you want to expose NoServiceVolumesFoundError?

**Review by pcrespov — COMMENTED** [review]
IN this PR you update `agent` service. The most important changes is the new rpc API which replaces the former rest API. The structure of the service is also updated

I have some comments and questions. thx

File: packages/service-library/src/servicelib/rabbitmq/rpc_interfaces/agent/volumes.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        RPCNamespace.from_entries(
+            {
+                "service": "agent",
+                "docker_node_id": docker_node_id,
+                "swarm_stack_name": swarm_stack_name,
+            }
+        ),
+        parse_obj_as(RPCMethodName, "remove_volumes_without_backup_for_service"),
+        node_id=node_id,
+        timeout_s=int(timedelta(minutes=60).total_seconds()),
```
Comment (pcrespov) [inline packages/service-library/src/servicelib/rabbitmq/rpc_interfaces/agent/volumes.py:31]: this is an interesting way of computing one minute in seconds every single time the call is done :-) 
Even if you want to keep the "algorithm", perhaps you can add it as a constant at the module level ?

File: services/docker-compose.yml [hunk 6]
```diff
@@ -1005,6 +1005,12 @@ services:
       AGENT_VOLUMES_CLEANUP_S3_SECRET_KEY: ${AGENT_VOLUMES_CLEANUP_S3_SECRET_KEY}
       AGENT_VOLUMES_CLEANUP_S3_BUCKET: ${AGENT_VOLUMES_CLEANUP_S3_BUCKET}
       AGENT_VOLUMES_CLEANUP_S3_PROVIDER: ${AGENT_VOLUMES_CLEANUP_S3_PROVIDER}
+      AGENT_DOCKER_NODE_ID: "{{.Node.ID}}"
```
Comment (YuryHrytsuk) [inline services/docker-compose.yml:1008]: Are you sure this is going to work?

**Comment by pcrespov** [comment]
This PR will also solve https://github.com/ITISFoundation/osparc-simcore/security/dependabot/1730

**Review by matusdrobuliak66 — APPROVED** [review]
Looks good from my side :+1:

File: services/agent/src/simcore_service_agent/core/api/rpc/_volumes.py (hunk 2 above)
Comment (GitHK) [reply services/agent/src/simcore_service_agent/core/api/rpc/_volumes.py:22]: Thanks for reminding me about it. Correct I have to expose it and use it when calling this interface.

**Review by pcrespov — APPROVED** [review]
thx

**Review by YuryHrytsuk — APPROVED** [review]

**Comment by GitHK** [comment]
Volume removal is working as expected, for autoscaled nodes and classic nodes.
`director-v2` is able to remove volumes from autoscaled nodes, since the `agent` service starts as soon as the node is `active`.

**Pull request merged**